design/20210302.certificate-request-policy.md

```diff
@@ -0,0 +1,509 @@
+}
+
+type CertificateRequestPolicyConditionType string
+
+const (
+    CertificateRequestPolicyConditionReady CertificateRequestPolicyConditionType = "Ready"
+)
+```
+
+Each non-enum string value can match using glob. This means that "*.example.com"
+can match with "foo.example.com", but not "example.com". We may decide that we
+would rather use regex, which would allow the user to express things _not_
+allowed ('^'), however regex can be complicated to write and is prone to errors
+(escaping full stops etc.).
+
+An empty or nil field can be considered as being "allow all" for that field.
+
+### Behaviour
+
+In order to keep backwards compatibility, and be in-line with the behaviour of
+existing resources such as network policies, the default action will be to
+"allow all" if no policy resource exists. As soon as a resource exists,
+cert-manager will "deny all" unless the CertificateRequest is accepted by the
+existing policy.
+
+The cert-manager CertificateRequest policy controller will watch
+CertificateRequest resources, which when they are created, will evaluate whether
+they pass approval. This is done by checking that a CertificateRequestPolicy
+which approves this request is [bound](#rbac) to the user who appears in the
+user info field of the request. If *any* CertificateRequestPolicy succeeds, the
+CertificateRequest is marked as *approved*. If *no* CertificateRequestPolicy
+succeeds, the CertificateRequest is marked as *denied*.
+
+### RBAC
+
+
+The binding of CertificateRequestPolicy and CertificateRequests is achieved
+using RBAC. This offers two dimensions to how the policy can be applied-
+namespace and user. A policy can be bound to an entire namespace regardless of
+user (system:authenticated), bound to particular users, or a mixture of both.
+
+When the policy controller evaluates whether a particular policy should be
+checked against a CertificateRequest, it will perform a SubjectAccessReview
+against the user present on the CertificateRequest. Since the
+CertificateRequestPolicy can be bound by both namespace and cluster scope, the
+SubjectAccessReview will be requested for both the namespace the
+CertificateRequest resides in, as well as at the cluster scope. The namespace
+scope will be checked first in an effort to improve performance (a large number
+of CertificateRequestPolicy resources are likely scoped by namespace). The
+SubjectAccessReview will be requested for every single CertificateRequestPolicy
+that exists, until and if a CertificateRequestPolicy which is bound approves the
+request. If no CertificateRequestPolicy approves the request, the
+CertificateRequest is marked as denied.
+
+The SubjectAccessReview will request for the attributes:
+
+```yaml
+Group: policy.cert-manager.io
+Verb: use
+Resource: CertificateRequestPolicy
+Name: <Name of the CertificateRequestPolicy being checked>
+Namespace: <Namespace of the CertificateRequest, empty if checking cluster scope>
+```
+
+---
+
+## External Policy Server
```

Comment:
What's the use-case for being able to extend this policy engine with external input?

Reply:
From what I understand, the purpose would be to be able to have use-cases such as:

"I want to enforce that the DNS name given in the DNSNames field matches the service name"

(example given by @james-w)

I guess this specific use-case would require the Policy Server request to contain not only the information about the certificate being requested but also a reference to the Service object... not sure if I understood this example though

Reply:
Does it make sense to you? @munnerz